src/rdkafka_cgrp.c

```diff
@@ -3169,9 +4158,270 @@ rd_kafka_cgrp_max_poll_interval_check_tmr_cb (rd_kafka_timers_t *rkts,
+                int j;
+
+                for (j=0; j<rkcg->rkcg_assignment->cnt; j++) {
+                        const rd_kafka_topic_partition_t *par =
+                                &rkcg->rkcg_assignment->elems[j];
+                        if (!strcmp(par->topic,
+                                    unsubscribing->elems[i].topic)) {
+                                rd_kafka_topic_partition_list_add(
+                                                        revoking,
+                                                        par->topic,
+                                                        par->partition);
+                                continue;
+                        }
+                }
+        }
+
+        if (revoking->cnt == 0) {
+                rd_kafka_topic_partition_list_destroy(revoking);
+                revoking = NULL;
+        }
+
+        return revoking;
+}
+
+
+/**
+ * @brief Handle a new subscription that is modifying an existing subscription
+ *        in the COOPERATIVE case.
+ */
+static rd_kafka_resp_err_t
+rd_kafka_cgrp_modify_subscription (rd_kafka_cgrp_t *rkcg,
+                                   rd_kafka_topic_partition_list_t
+                                   *rktparlist) {
+        rd_kafka_topic_partition_list_t *unsubscribing_topics;
+        rd_kafka_topic_partition_list_t *revoking;
+        rd_list_t *tinfos;
+        rd_kafka_topic_partition_list_t *errored;
+        int metadata_age;
+
+        rkcg->rkcg_flags &= ~RD_KAFKA_CGRP_F_WILDCARD_SUBSCRIPTION;
+
+        if (rd_kafka_topic_partition_list_regex_cnt(rktparlist) > 0)
+                rkcg->rkcg_flags |= RD_KAFKA_CGRP_F_WILDCARD_SUBSCRIPTION;
+
+        /* Topics in rkcg_subscribed_topics that don't match any pattern in
+           the new subscription. */
+        unsubscribing_topics = rd_kafka_cgrp_get_unsubscribing_topics(
+                                        rkcg, rktparlist);
+
+        /* Currently assigned topic partitions to that are no longer
+           desired. */
+        revoking = rd_kafka_cgrp_calculate_subscribe_revoking_partitions(
+                                        rkcg, unsubscribing_topics);
+
+        if (unsubscribing_topics)
+                rd_kafka_topic_partition_list_destroy(unsubscribing_topics);
+
+        rd_kafka_topic_partition_list_destroy(rkcg->rkcg_subscription);
+        rkcg->rkcg_subscription = rktparlist;
+
+        if (rd_kafka_cgrp_metadata_refresh(rkcg, &metadata_age,
+                                           "modify subscription") == 1) {
+                rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER, "JOIN",
+                             "Group \"%.*s\": postponing join until "
+                             "up-to-date metadata is available",
+                             RD_KAFKAP_STR_PR(rkcg->rkcg_group_id));
+
+                /* Join will occur after metadata refresh completes */
+                return RD_KAFKA_RESP_ERR_NO_ERROR;
```

Comment:
Memory leaks here, is there a test that covers this case where metadata needs to be refresh on subscription update, i.e., when a new topic is added perhaps?

Reply:
yes, more than one test. i guess I haven't run with asan since

Reply:
noted